xds/src/main/java/io/grpc/xds/SharedCallCounterMap.java

```diff
@@ -52,19 +52,28 @@ static SharedCallCounterMap getInstance() {
 
   @Override
   public synchronized AtomicLong getOrCreate(String cluster, @Nullable String edsServiceName) {
+    AtomicLong counter = null;
     Map<String, CounterReference> clusterCounters = counters.get(cluster);
+    if (clusterCounters != null) {
+      CounterReference ref = clusterCounters.get(edsServiceName);
+      if (ref != null) {
+        counter = ref.get();
+      }
+    }
+    // In the case of ref.get() == null, must call cleanQueue() prior to creating a new map entry,
+    // otherwise the new entry will be deleted by any later cleanQueue().
```

Comment:
Seems the reliable fix for this would be to check that the `CounterReference` instance is the current value in `counters` before doing `remove()`.

Reply:
Good idea!